### PR TITLE
PLT-7407: Back-end plugins

### PR DESCRIPTION
**Pull request by ccbrown**
#### Summary
It is finished.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-7407?filter=14400

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

File: api4/api.go [hunk 1]
```diff
@@ -150,7 +150,7 @@ func InitApi(full bool) {
 	BaseRoutes.PublicFile = BaseRoutes.Root.PathPrefix("/files/{file_id:[A-Za-z0-9]+}/public").Subrouter()
 
 	BaseRoutes.Plugins = BaseRoutes.ApiRoot.PathPrefix("/plugins").Subrouter()
-	BaseRoutes.Plugin = BaseRoutes.Plugins.PathPrefix("/{plugin_id:[A-Za-z0-9\\_\\-]+}").Subrouter()
+	BaseRoutes.Plugin = BaseRoutes.Plugins.PathPrefix("/{plugin_id:[A-Za-z0-9\\_\\-\\.]+}").Subrouter()
```
Comment (ccbrown) [inline api4/api.go]: Proposing we allow dots to facilitate reverse-dns notation for plugin ids.

File: app/plugins.go [hunk 2]
```diff
@@ -19,63 +19,96 @@ import (
 	"github.com/mattermost/mattermost-server/model"
 	"github.com/mattermost/mattermost-server/utils"
 
-	"github.com/mattermost/mattermost-server/app/plugin"
+	builtinplugin "github.com/mattermost/mattermost-server/app/plugin"
```
Comment (ccbrown) [inline app/plugins.go]: Relabeling the pre-existing stuff as "built-in" plugins. Once we port the JIRA / LDAP stuff, we can get rid of this altogether.

File: utils/extract.go [hunk 3]
```diff
@@ -61,7 +61,7 @@ func ExtractTarGz(gzipStream io.Reader, dst string) ([]string, error) {
 				return nil, fmt.Errorf("ExtractTarGz: MkdirAll() failed: %s", err.Error())
 			}
 
-			outFile, err := os.Create(path)
+			outFile, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE|os.O_TRUNC, os.FileMode(header.Mode))
```
Comment (ccbrown) [inline utils/extract.go]: Gotta preserve permissions for executables.

File: plugin/pluginenv/environment.go [hunk 4]
```diff
@@ -200,13 +202,54 @@ func (env *Environment) Shutdown() (errs []error) {
...
+
+// ServeHTTP invokes the ServeHTTP hook for the plugin identified by the request or responds with a
+// 404 not found.
+//
+// It expected the plugin_id mux parameter to be set.
+func (h *EnvironmentHooks) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	params := mux.Vars(r)
+	if id, ok := params["plugin_id"]; ok {
+		h.env.mutex.Lock()
+		defer h.env.mutex.Unlock()
```
Comment (ccbrown) [inline plugin/pluginenv/environment.go]: TODO: Concurrency. :P This lock shouldn't be too hard to optimize out, and I'm pretty positive we'll want to do that, but I'm putting it off until later.

**Comment by coreyhulen** [comment]
Somewhat unrelated but I wonder if we should allow a system admin the ability to disable plugin uploads.  So for the ultra paranoid you can only manually install plugins by copying the files to server yourself.

File: plugin/pluginenv/environment.go (hunk 4 above)
Comment (ccbrown) [reply plugin/pluginenv/environment.go]: Made a ticket for this: https://mattermost.atlassian.net/browse/PLT-7572

**Comment by ccbrown** [comment]
Created a task to further expand the API:

https://mattermost.atlassian.net/browse/PLT-7608

And made a really basic repo if anyone wants to actually see a "hello world" plugin:

https://github.com/ccbrown/mattermost-plugin-demo

Once we have a bit more of an API, maybe I can make it into a cooler demo.

File: app/plugins.go [hunk 5]
```diff
@@ -176,48 +211,43 @@ func (a *App) ActivatePlugins() {
...
+	if err != nil {
+		return nil, model.NewAppError("UnpackAndActivatePlugin", "app.plugin.filesystem.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 
-	splitPath := strings.Split(filenames[0], string(os.PathSeparator))
-
-	if len(splitPath) == 0 {
-		return nil, model.NewAppError("UnpackAndActivatePlugin", "app.plugin.bad_path.app_error", nil, err.Error(), http.StatusBadRequest)
+	if len(dir) == 1 && dir[0].IsDir() {
+		tmpPluginDir = filepath.Join(tmpPluginDir, dir[0].Name())
```
Comment (ccbrown) [inline app/plugins.go]: Changed this up a little to allow tarballs where files are placed directly into the root. For example, this should work:

`tar -czvf mattermost-plugin-demo.tar.gz mattermost-plugin-demo plugin.yaml`

File: plugin/rpcplugin/io.go [hunk 6]
```diff
@@ -2,26 +2,169 @@ package rpcplugin
...
+	}
+}
+
+func (w *asyncWriteCloser) Close() error {
+	w.closeOnce.Do(func() {
+		close(w.close)
+	})
+	return nil
+}
+
```
Comment (ccbrown) [inline plugin/rpcplugin/io.go]: Eliminating a race condition. Hoping for Go 1.10 we'll get a way to safely interrupt or time out `(*os.File).Read`, simplifying all of this. (See https://github.com/golang/go/issues/18507)

**Comment by ccbrown** [comment]
@coreyhulen The option to disable plugin uploads might be a good idea for our open-to-world test servers. Of course, another good idea might be to require VPN for those. :-/

**Pull request merged**